e2e/nomostest/clean.go

```diff
@@ -218,15 +220,66 @@ func Clean(nt *NT, failOnError FailOnError) {
 		nt.T.Fatal("error waiting for test objects to be deleted")
 	}
 
-	DeleteManagedNamespaces(nt)
+	if err := DeleteManagedNamespaces(nt); err != nil {
+		if failOnError {
+			nt.T.Fatal(err)
+		} else {
+			nt.T.Error(err)
+		}
+	}
 	deleteImplicitNamespaces(nt, failOnError)
 }
 
 // DeleteManagedNamespaces deletes all the namespaces managed by Config Sync.
-func DeleteManagedNamespaces(nt *NT) {
-	nt.T.Logf("Started deleting managed namespaces at %v", time.Now())
-	nt.MustKubectl("delete", "ns", "-l", fmt.Sprintf("%s=%s", metadata.ManagedByKey, metadata.ManagedByValue))
-	nt.T.Logf("Finished deleting managed namespaces at %v", time.Now())
+func DeleteManagedNamespaces(nt *NT) error {
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		nt.T.Logf("[CLEANUP] Deleting managed namespaces took %v", elapsed)
+	}()
+
+	// Delete all Namespaces with the test label (except protected).
+	nsList := &corev1.NamespaceList{}
+	if err := nt.List(nsList, withLabelListOption(metadata.ManagedByKey, metadata.ManagedByValue)); err != nil {
+		return err
+	}
+
+	// Error if any protected namespace are managed by config sync
+	// (managed-by label added) and not reverted by the test.
+	protectedNamespacesWithTestLabel, nsListItems := filterNamespaces(nsList.Items,
+		protectedNamespaces...)
+	if len(protectedNamespacesWithTestLabel) > 0 {
+		return errors.Errorf("protected namespace(s) managed by config sync and cannot be cleaned: %+v",
+			protectedNamespacesWithTestLabel)
+	}
+	return DeleteNamespaces(nt, nsListItems)
+}
+
+// DeleteNamespaces cleans up one or more Namespaces and all their namespaced objects.
+// Use this for bulk deletion of Namespaces.
+func DeleteNamespaces(nt *NT, nsList []corev1.Namespace) error {
+	if len(nsList) == 0 {
+		return nil
+	}
+	nt.T.Logf("[CLEANUP] Deleting Namespaces (%d)", len(nsList))
```

Comment:
Consider moving this line above the zero length check so we know if the array was zero length from the test logs

Reply:
Done